doggos/inference/mamdani_inference_system.py

```diff
@@ -1,6 +1,112 @@
+    --------------------------------------------
+    __rule_base: List[Rule]
+        fuzzy knowledge base used for inference
+
+    Methods
+    --------------------------------------------
+    infer(self, features: Dict[Clause, List[MembershipDegree]], method: str = 'karnik_mendel') -> float:
+        infer decision from knowledge base
+
+    Examples:
+    --------------------------------------------
+    Creating simple mamdani inference system and infering decision
+    >>> rules = [first_rule, second_rule, third_rule]
+    >>> features: Dict[Clause, MembershipDegree] = fuzzifier.fuzzify(dataset)
+    >>> mamdani = MamdaniInferenceSystem(rules)
+    >>> mamdani.output(features, 'karnik_mendel')
+    0.5
+    """
+    __rule_base: List[Rule]
+
+    def __init__(self, rules: List[Rule]):
+        """
+        Create mamdani inference system with given knowledge base
+        All rules should have the same consequent type and consequents should be defined on the same domain
+        :param rules: fuzzy knowledge base used for inference
+        """
+        super().__init__(rules)
+        self.__validate_consequents()
+
+    def infer(self, defuzzification_method: Callable, features: Dict[Clause, List[MembershipDegree]]) \
+            -> Iterable[float]:
+        """
+        Inferences output based on features of given object using chosen method
+        :param defuzzification_method: 'KM', 'COG', 'LOM', 'MOM', 'SOM', 'MeOM', 'COS'
+        :param features: dictionary of linguistic variables and their values
+        :return: decision value
+        """
+        if not isinstance(features, Dict):
+            raise ValueError("Features must be fuzzified dictionary")
+        if not isinstance(defuzzification_method, Callable):
+            raise ValueError("Defuzzifiaction method must be callable")
+
+        values = np.array(features.values())
+        degrees = values[0]
+
+        for i in range(len(degrees)):
+            single_features = {}
+            for clause, memberships in features.items():
+                single_features[clause] = memberships[i]
+
+            if self.__is_consequent_type1():
+                domain, membership_functions = self.__get_domain_and_membership_functions(single_features)
+                yield defuzzification_method(domain, membership_functions)
+            else:
+                domain, lmfs, umfs = self.__get_domain_and_memberships_for_it2(single_features)
+                yield defuzzification_method(lmfs, umfs, domain)
+
+    def __validate_consequents(self):
+        for rule in self._rule_base:
+            if not isinstance(rule.consequent, MamdaniConsequent):
+                raise ValueError("All rule consequents must be mamdani consequents")
+
+    def __is_consequent_type1(self):
+        return isinstance(self.__rule_base[0].consequent.clause.fuzzy_set, Type1FuzzySet)
+
+    def __get_domain_and_memberships_for_it2(self, features: Dict[Clause, List[MembershipDegree]]) \
```

Comment:
Consider name:
__get_domain_and_outputs_for_it2

Reply:
Would rather change output variables inside of functions to membership_function - provides better description